File: PHP/linked_list.php
<?php

class Node {
    public $data;
    public $next;

    public function __construct($data, $next=NULL) {
        $this->data = $data;
        $this->next = $next;
    }
}

class LinkedList {
    private $head;

    public function __construct($data, $next=NULL) {
        $this->head = new Node($data, $next);
    }

    public function insert($data) {
        $node = new Node($data);
        if (empty($this->head)) {
            $this->head = $node;
            return $this->head;
        }
        $node->next = $this->head;
        $this->head = $node;
        return $this->head;
    }

    public function delete($data) {
        $node = $this->search($data);
        if (empty($this->head) || empty($node)) {
            return False;
        } elseif ($this->head == $node) {
            $this->head = $node->next;
            unset($node);
            return True;
        }
        $prev = $this->getPrevNode($node);
        $prev->next = $node->next;
        unset($node);
        return True;
    }

    public function search($data) {
        if (empty($this->head)) {
            return;
        }
        $curr = $this->head;
        while ($curr != NULL) {
            if ($curr->data == $data) {
                return $curr;
            }
            $curr = $curr->next;
        }
        return NULL;
    }

    public function getHead() {
        return $this->head;
    }

    public function printList() {
        if (empty($this->head)) {
            return;
        }
        $curr = $this->head;
        while ($curr != NULL) {
            echo $curr->data . " ";
            $curr = $curr->next;
        }
        echo PHP_EOL;
    }

    public function getPrevNode(&$node) {
        if (empty($this->head)) {
            return;
        }
        $prev = NULL;
        $curr = $this->head;
        while (!empty($curr)) {
            if ($curr->data == $node->data) {
                return $prev;
            }
            $prev = $curr;
            $curr = $curr->next;
        }
        return NULL;
    }

    public function reverse() {
        if (empty($this->head)) {
            return;
        }
        $prev = NULL;
        $curr = $this->head;
        while (!empty($curr)) {
            $temp = $curr->next;
            $curr->next = $prev;
            $prev = $curr;
            $curr = $temp;
        }
        $this->head = $prev;
    }
}

// $list = new LinkedList(1);
// $list->insert(2);
// $list->insert(3);
// $list->insert(4);
// $list->printList();
// $list->reverse();
// $list->printList();

File: JAVA/LinkedList.java
import java.util.*;
import Node;

public class LinkedList {
	public LinkedList(int data) {
		Node first = new Node(data);		
	}

	public empty() {
		if (this.first.next == null) {
			return true;
		} else {
			return false;
		}
	}

	public void insert(int data, Node pos) {
		Node temp = new Node(data);
		temp.next = pos.next;
		pos.next = temp;
	}

	public void delete(Node pos) {
		//Deletes the node after the node referenced by pos
		pos.next = pos.next.next;
	}

	public void print() {
		boolean isEmpty = true;

		while (this.first.next != null) {			
			System.out.println(this.first.data);
			this.first = this.first.next;
			isEmpty = false;
		}
		if (this.first.next == null) {
			if (isEmpty) {
				System.out.println("List is empty.");
			} else {
				System.out.println(this.first.data);
			}			
		}
	}

	public static int main(String[] args) {
		LinkedList list1 = new LinkedList(1);
		list1.print();
		list1.insert(2, LinkedList.first);
		list1.print();
	}
}